### ISSUE TITLE
Mongo update error after logging into Facebook

### ISSUE THREAD

**Issue by node-monk** [feathersjs/feathers-authentication#244]
Sorry in advance for all the code posted in this, but I almost have the Facebook login working. It worked once and created a user in the mongo database, but after that it was throwing these errors in the debugger and sending me to /auth/failure. 

In my setup, I have feathers as an api subfolder and its access by the '/api/v1' url so I have to do some redirection in /index.js for the auth/facebook routes.

```
  feathers-authentication:oauth2 Updating user: 578c46ef0a0d2d550e16aea7 +12s
  feathers-authentication:middleware An authentication error occurred. +8ms [Error: Invalid atomic update value for $__original_save. Expected an object, received function]
```

/api/config/default.json

``` js
{
  "host": "localhost",
  "port": 8080,
  "mongodb": "mongodb://localhost:27017/nnbbApi",
  "public": "../public/",
  "auth": {
    "shouldSetupSuccessRoute": false,
    "token": {
      "secret": "..."
    },
    "facebook": {
      "clientID": "...",
      "clientSecret": "...",
      "permissions": {
        "scope": ["public_profile", "email"]
      },
      "profileFields": ["id", "displayName", "photos", "email", "first_name", "last_name", "age_range"]
    },
    "local": {}
  }
}
```

/api/middleware/auth

``` js
'use strict';

const handler = require('feathers-errors/handler');
const notFound = require('./not-found-handler');
const logger = require('./logger');

module.exports = function() {
  // Add your custom middleware here. Remember, that
  // just like Express the order matters, so error
  // handling middleware should go last.
  const app = this;

  app.get('/auth/success', function(req, res) {
    res.redirect('/member/profile');
  });

  app.use(notFound());
  app.use(logger(app));
  app.use(handler());
};

```

/api/services/auth/index.js

``` js
'use strict';

const authentication = require('feathers-authentication');
const FacebookStrategy = require('passport-facebook').Strategy;


module.exports = function() {
  const app = this;

  let config = app.get('auth');
  config.facebook.strategy = FacebookStrategy;

  app.set('auth', config);

  app.configure(authentication(config));
};

```

``` js

'use strict';

// user-model.js - A mongoose model
//
// See http://mongoosejs.com/docs/models.html
// for more of what you can do here.

const mongoose = require('mongoose');
const Schema = mongoose.Schema;

const userSchema = new Schema({
  facebookId: {
    type: String
  },
  facebook: {
    type: Schema.Types.Mixed
  },
  // email: {
  //   type: String,
  //   required: true,
  //   unique: true
  // },
  // password: {
  //   type: String,
  //   required: true
  // },
  // Added in by Michael Watson
  profile: {
    firstname: {
      type: String
    },
    lastname: {
      type: String
    },
    locationName: {
      type: String
    },
    bio: {
      type: String
    },
    location: {
      type: {
        type: String,
        default: 'Point'
      },
      coordinates: [Number]
    },
    subscriptions: [{
      subscriptionId: {
        type: Schema.Types.ObjectId,
        ref: 'subscription'
      },
      startDate: {
        type: Date
      },
      endDate: {
        type: Date
      },
    }],
    photo: {
      type: String
    }
  },

  createdAt: {
    type: Date,
    'default': Date.now
  },
  updatedAt: {
    type: Date,
    'default': Date.now
  }
});

const userModel = mongoose.model('user', userSchema);

module.exports = userModel;

```

/start.js

``` js
var express = require('express');
var path = require('path');

/**
 * Pull in Sub Application
 */
var feathers = require('feathers');
var api_application = require('./api/src/app');
var api = feathers().use('/', api_application);
var app = express().use('/api/v1', api);
var querystring = require('querystring');
/**
 * Load environment configuration
 */
var config = require(path.join(__dirname, 'application', 'configs', `${process.env.NODE_ENV}`));
try {
  require(path.join(__dirname, 'application', 'setup'))({
    app: app,
    config: config,
    api: api_application
  });
} catch (e) {
  console.error(e.stack);
  process.exit();
}

app.get('/auth/facebook', (req, res) => {
  console.log('rerouting to auth/facebook');
  res.redirect('/api/v1/auth/facebook');
});

app.get('/auth/success', (req, res) => {
  console.log('rerouting to api success.');
  res.redirect('/api/v1/auth/success');
});

app.get('/auth/failure', (req, res) => {
  res.redirect('/api/v1/auth/failure');
});

app.get('/auth/facebook/callback', (req, res) => {
  console.log('rerouting to callback.', req.query);

  res.redirect(`/api/v1/auth/facebook/callback?${querystring.stringify(req.query)}`);
});


/**
 * Start server
 */
const PORT = process.env.PORT || 8080;
var server = app.listen(PORT, () => {
  console.log(`NaNannyBooBoo Listening or port ${PORT}`);
});
api.setup(server);

```

**Comment by ekryski** [comment]
Hmm. That's no good. We have a pretty big update coming pretty soon that may fix this (or at least will make it easier to debug).

If you can provide a link to the whole project that would be much easier for us to help debug. Nothing is standing out right away other than all the redirection seems like it might not be necessary.

Not that I think this is your issue but you just do:

```js
var api = require('./api/src/app');
var api = feathers().use('/api/v1', api_application);
```

No need to include express because Feathers **is** Express under the hood. If you haven't already you can check out [feathers-demos](https://github.com/feathersjs/feathers-demos/tree/master/examples) to see some examples of sub-apps and auth.

Auth with sub-apps doesn't really work very well. It's typically a bit better if you have your auth at the top level.

**Comment by node-monk** [comment]
Thanks for looking into this. I'll keep messing with it and maybe your next update will resolve it. There is not currently a url to access the project, it's still in early development on my local machine. Feathers sure has made it easy however to get up and running, and using service from the client ROCKS!!! Keep up the good work!!

**Comment by node-monk** [comment]
I think I figured it out. Looks like Mongoose was adding $__original_save to the data before it was updating. The update rejected $__original_save because it was a function -- not an object. 
In Feathers-Mongoose (line 247) I solved the problem by wrapping "data" with "Object.keys(data)" as shown below 
```js
        return this.Model.update(params.query, Object.keys(data), options).lean(this.lean).exec().then(function () {

```

I'll go do a pull request for this change in feathers-mongoose.

**Comment by node-monk** [comment]
Hope this helps, I was doing some digging and the issue is in the Patch method of feathers-mongoose/service.js (line 247)
```js
return this.Model.update(params.query, data, options).lean(this.lean).exec().then(function () {

```
The "data" object getting passed into the update has an $__original_save property thats breaking the update because its a function not an object.

I made a simple update by having it pass the "data._doc" and the update didn't throw an error(and logged me in), but I am not sure if this is the correct solution or not. Below is the change I made

```js
return this.Model.update(params.query, data._doc, options).lean(this.lean).exec().then(function () {


```

**Comment by rohit-romley** [comment]
Thanks a lot for the solution man.

**Comment by ekryski** [comment]
Thanks for doing that digging @node-monk! Looks like `toObject` is not getting called or something to that effect. Going to leave this open because I _think_ this is a legit issue.

As a side note when you initialize a Mongoose service to can set `{lean: true}` and all your queries will be lean so you won't need to do what you're doing.

**Comment by simonjoom** [comment]
hi ,
I ve got the same issue 
// var data = Object.assign({}, user, data);
the problem come from Object.assign, i guess it's better to create a new object without link to old reference (else data contain some $__original_save $__original_remove who do the problem) 

//me i used that 👍 and no problem anymore
var data = JSON.parse(JSON.stringify(data ));
delete data._id;
  delete data.__v; 

          return app.service(options.userEndpoint).patch(id, data).then(function (updatedUser) {

**Comment by ekryski** [comment]
This should be resolved with Auth v1.0.0. Please the [migration guide](https://github.com/feathersjs/feathers-authentication/blob/master/docs/migrating.md).

If not, let's re-open and I will dive deeper but I haven't been able to reproduce on the new version.
